ENH: allow apps to set a csv path for a model within the command.

bcipolli opened the pull request:
`load_*` allow passing `csv_path`, but `load` does not. However, `load` does a lot of useful database manipulation. To use it, `csv_path` must be specified on the model, but sometimes it's more convenient to have it localized to the command itself (which frees it do the downloading and loading at any location).

This PR allows `csv_path` to be passed to `load`.

An alternate config would be to refactor all the database manipulation into a separate function, and call it from `load`.

bcipolli commented on calaccess_raw/management/commands/loadcalaccessrawfile.py:
```diff
@@ -60,7 +60,7 @@ def load(self, model_name):
             'django.contrib.gis.db.backends.postgis'
-                ):
+        ):
             self.load_postgresql(model, csv_path)
```
PEP8 change that my editor chose to make ... I can back this out if needed.

palewire commented:
Open to this idea, though I'm unclear on the use case. Can you let us know why you are itching for this feature so we can better understand where you're coming and perhaps draft a more comprehensive refactor.

bcipolli commented:
I have models that don't directly associate to CSV data, but I want to load them from a CSV file. I manipulate the data into the format of my models, then save.

So I'd rather keep the CSV info in the command; it feels strange to define on the models.

I think this PR is consistent with how parameters are already passed in the command. I don't see the need to harder work than this, but I'm open to it if you feel there's something more general.

palewire commented:
Interesting. More than anything, I'm fascinated by how people are using our code. So you're this command to load models and CSVs separate from those in the CAL-ACCESS database download?

bcipolli commented:
I looked a little more closely at this, and I believe the better use-case to discuss is: we are pulling `netfile` data, and there are multiple CSV files we are pulling, per model (based on the city "shortcut" name and the year). So `get_csv_path` would have to take parameters.

Yes, we are mirroring the `calaccess_raw` way of doing things when we access `netfile` data.

palewire commented:
I see. That's interesting to learn. This seems like a totally reasonable option to allow an override on, IMHO. My only little design nit to pick is that I feel like any option we introduce into the Python API should also be surfaced on the ``manage.py`` command-line interface. How would you feel about a modification to this patch that would devolve the current keyword argument into the ``handle`` function where it would be connected to a new ``make_option``?

bcipolli commented:
I would be fine with that. 

Just as a note, that doesn't make sense for us--we iterate over cities in our command and compute the `csv_path` on the fly. So in our command, I'd have to find a way to strip out that option, or at least error if a user used it.

palewire commented:
I'm imagining a slightly modified API that would offer two rough entry points. Here is some psuedocode to try to articulate the idea.

Current, the command-line interfaces requires the model name as an argument, but infers the CSV automatically.
```
$ python manage.py loadcalaccessrawfile RcptCd
```

That could be expanded to allow the CSV to be specified as an option.
```
$ python manage.py loadcalaccessrawfile RcptCd --csv=/home/ben/my-data.csv
```

That same option could then be accessed in Python scripts via [the textbook Django method](https://docs.djangoproject.com/en/1.9/ref/django-admin/#running-management-commands-from-your-code).

```
from django.core import management
management.call_command('loadcalaccessrawfile', 'RcptCd', csv="/home/ben/my-data.csv")
```

Sound okay?

bcipolli commented:
Right. I understood that well. That's fine by me. I'm just telling you, that interface doesn't fit our use-case, and we'll have to disable it.

For us, a single model can have multiple csv files, depending on how many cities/years of data you choose to iterate over. The csv files must be computed. I guess we could accept a format string for the csv, or a directory path... but just was pointing out earlier that the interface you're proposing isn't a clean fit for our use-case.

Happy to implement it nonetheless.

palewire commented:
Okay. Glad I'm making some sense! Though I think I'm a little dense and missing how we'd be cutting you out with that modification. If you have, for instance, three CSVs being inserted into a single table, would this model let you do something like this?

```python
from django.core import management
management.call_command('loadcalaccessrawfile', 'NetFileTable', csv="/home/ben/one.csv")
management.call_command('loadcalaccessrawfile', 'NetFileTable', csv="/home/ben/two.csv")
management.call_command('loadcalaccessrawfile', 'NetFileTable', csv="/home/ben/three.csv")
```

bcipolli commented:
Oh, so sorry for being unclear. We are overriding the `load` method, and iterating within the command itself. It is within the command that we download the list of cities available, and then filter and loop.

So this isn't how we'd use the command. We'd be calling these functions within the command itself:
https://github.com/caciviclab/disclosure-backend/blob/master/netfile_raw/management/commands/downloadnetfilerawdata.py

palewire commented:
Ah. Now I get it! 

This is obviously pretty far down the Django pedant rabbit hole, but wouldn't it work to substitute out a line like this

```python
 super(Command, self).load(
            'NetFileAgency', csv_path=self.agency_csv_path)
```

with this

```python
management.call_command('loadcalaccessrawfile', 'NetFileAgency', csv="/home/ben/three.csv")
```
If there's something obvious I'm missing about that thank you in advance for your patience.

bcipolli commented:
That doesn't make sense to me because you're calling `call_command` from within a subclass of that command. When subclassing, you should have access to the member functions, and call them directly. It strikes me as a very odd way to access member functions of a superclass.

So, if we were to do that, I'd rather not subclass the `loadcalaccessrawfile` command at all. Then it reads correctly--to call external commands, use `call_command`. But we want to subclass, because we reuse most of the internal machinery of `loadcalaccessrawfile`.

Is there a downside to the way we're using it? I thought subclassing the command was very legitimate, and the functions aren't marked as private (with a leading underscore), so legitimate to work with those directly as well.

palewire commented:
Ah. I see your point. I wonder if we can thread the needle and get both things if a line [like this](https://github.com/bcipolli/django-calaccess-raw-data/blob/explicit-csv/calaccess_raw/management/commands/loadcalaccessrawfile.py#L41) in the current command
```python
self.load(label)
```
Is slightly expanded to follow the retrieval of the model pull the option from the CLI and handle command (if it exists) and then passes it into the load function.

```python
csv_path = options.get("csv") # <-- This will return None unless the user specifies with CLI/call_command
self.load(label, csv_path=csv_path)
```

If a None is passed in, it can continue to pull the path from the model. If the CLI or call_command gets a keyword argument with an alternative path, it will get passed along. And if you pass in your own keyword argument directly to load I think it might work too.

bcipolli commented:
@palewire Yes, that's exactly what I was thinking: keep the `load` interface as I've done here, and simply add on the `csv` command-line arg as you showed. It will be easy, seems clean.

Just, we will have to disable that argument from our command, because it doesn't really make sense... I feel like I should be quiet about it.

palewire commented:
Ooooooh okay. I think I finally get it! If we add the option to the base command your command will inherit it and then users might mistakenly use it when you don't want them to.

Sorry this took me so long to get. You guys have, in impressive form, run so far ahead of my thinking about the command that I'm just catching up. 

If I'm reading the Django development notes correctly, we might need to consider a broader refactor of how we configure management command options. In 1.8 they moved away from the ``make_option`` system we use now to [a new system](https://docs.djangoproject.com/en/1.9/howto/custom-management-commands/#accepting-optional-arguments). At some point they might deprecate the old system and we'd probably be smart to consider how we could rewrite the options we have to work in the new scheme, and perhaps solve the issue you've raised here along the way.

bcipolli commented:
@palewire I am up for looking at the refactoring you talk about, but I'd really appreciate to open a new issue for that and make the straightforward changes in front of us. We're pushing pretty hard on a project using the netfile data, and want to focus my efforts on that at the moment.

Completely agree with your sentiment, but would much prefer to deal with this down the road, since the remaining changes here are so lightweight and very consistent with other changes that need to be made.

bcipolli commented:
@palewire I added the command-line arg as requested, rebased to integrate the `database` change. Very simple change (2 mins). Hope we can submit this as-is and open an issue? :dancers:

palewire commented:
Thank you. We'll get this merged and create that new issue we discussed. Thank you for your patience and contributions.
